spring-xd-dirt/src/main/java/org/springframework/xd/dirt/module/SynchronizingModuleRegistry.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * <p>This is useful as reading
+ * Boot uber-jars requires local {@code java.io.File} access, but modules may reside in a remote registry (<i>e.g.</i>
+ * backed by HDFS). For such a case, simply use this registry as the main registry, configuring it with the {@code hdfs://} registry as
+ * its source and the {@code file://} one as its target.</p>
+ *
+ * @author Eric Bottard
+ */
+public class SynchronizingModuleRegistry implements WritableModuleRegistry {
+
+	private final WritableModuleRegistry sourceRegistry;
+
+	private final WritableModuleRegistry targetRegistry;
+
+	private final StalenessCheck staleness = new MD5StalenessCheck();
+
+	private ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
+
+	public SynchronizingModuleRegistry(WritableModuleRegistry sourceRegistry, WritableModuleRegistry targetRegistry) {
+		Assert.notNull(sourceRegistry, "sourceRegistry cannot be null");
+		Assert.notNull(targetRegistry, "targetRegistry cannot be null");
+		this.sourceRegistry = sourceRegistry;
+		this.targetRegistry = targetRegistry;
+	}
+
+
+	@Override
+	public ModuleDefinition findDefinition(String name, ModuleType moduleType) {
+		ModuleDefinition remoteDefinition = sourceRegistry.findDefinition(name, moduleType);
+		if (remoteDefinition == null || remoteDefinition.isComposed()) {
+			return remoteDefinition;
+		}
+
+		copyIfStale(remoteDefinition);
+		return currentTargetDefinition(remoteDefinition);
+	}
+
+	@Override
+	public List<ModuleDefinition> findDefinitions(String name) {
+		List<ModuleDefinition> remoteDefinitions = sourceRegistry.findDefinitions(name);
+		return refresh(remoteDefinitions);
+	}
+
+	@Override
+	public List<ModuleDefinition> findDefinitions(ModuleType type) {
+		List<ModuleDefinition> remoteDefinitions = sourceRegistry.findDefinitions(type);
+		return refresh(remoteDefinitions);
+	}
+
+	@Override
+	public List<ModuleDefinition> findDefinitions() {
+		List<ModuleDefinition> remoteDefinitions = sourceRegistry.findDefinitions();
+		return refresh(remoteDefinitions);
+	}
+
+	/**
+	 * Return a refreshed list of definitions from the target registry.
+	 */
+	private List<ModuleDefinition> refresh(List<ModuleDefinition> remoteDefinitions) {
+		List<ModuleDefinition> result = new ArrayList<ModuleDefinition>(remoteDefinitions.size());
+		for (ModuleDefinition remoteDefinition : remoteDefinitions) {
+			if (remoteDefinition.isComposed()) {
+				result.add(remoteDefinition);
+			}
+			else {
+				copyIfStale(remoteDefinition);
+				result.add(currentTargetDefinition(remoteDefinition));
+			}
+		}
+
+		return result;
+	}
+
+
+	/**
+	 * Checks whether the target version of the definition is stale, and if it is, triggers an update (as a deletion then
+	 * new registration).
+	 */
+	private synchronized void copyIfStale(ModuleDefinition remoteDefinition) {
+		SimpleModuleDefinition simpleRemoteDefinition = (SimpleModuleDefinition) remoteDefinition;
```

Comment:
We had this discussion way back then, and this is exactly what I said. You were against instanceof checks. But there are really only 3 ways to do this, IMHO:
1. instanceof checks at usage site
1.1. hiding the instanceof check behind a method (what we do here)
2. double dispatch (ie visitor), which seems overkill here as we only have 2 implementations
3. collection-style exposition, which is actually a variation of double dispatch in a way, ie

```
CompositeModuleDefinition {
Iterable<String> or Optional<String> getLocation() {return "none"}
}

SimpleModuleDefinition {
 Iterable<ModuleDefinition> getChildren() {return empty()}
}
```